python_modules/dagster/dagster/_core/utils.py

```diff
@@ -127,6 +128,35 @@ def parse_env_var(env_var_str: str) -> Tuple[str, str]:
 class InheritContextThreadPoolExecutor(ThreadPoolExecutor):
     """A ThreadPoolExecutor that copies over contextvars at submit time."""
 
+    def __init__(
+        self,
+        max_workers: int | None = None,
+        thread_name_prefix: str = "",
+        initializer: Callable[..., object] | None = None,
+        initargs: tuple[Any, ...] = ...,
+    ) -> None:
+        super().__init__(max_workers, thread_name_prefix, initializer, initargs)
+        self._futures = []
+
     def submit(self, fn, *args, **kwargs):
         ctx = copy_context()
-        return super().submit(ctx.run, fn, *args, **kwargs)
+        future = super().submit(ctx.run, fn, *args, **kwargs)
+        self._futures = [
+            future for future in self._futures if not future.done()
+        ]  # clean up done futures
+        self._futures.append(future)
+        return future
+
+    @property
+    def max_workers(self) -> int:
+        return self._max_workers
+
+    @property
+    def num_running_futures(self) -> int:
+        return (
+            len([future for future in self._futures if not future.done()]) - self.num_queued_futures
+        )
+
+    @property
+    def num_queued_futures(self) -> int:
+        return self._work_queue.qsize()
```

Comment:
hopefully these private members are around in all python versions (and those to come)

Reply:
yea I had the same thought when we initially started messing with this stuff. But I figure the refactor will at least be limited in scope if/when we need to make changes.